src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
+        message_bus: *MessageBus,
+        /// A value of -1 indicates that the connection is currently being
+        /// closed or is inactive.
+        fd: os.socket_t = -1,
+        /// The peer is determined by inspecting the first message header
+        /// received.
+        peer: union(enum) {
+            /// No peer is currently connected.
+            none: void,
+            /// A connection has been established but the first header has not yet been recieved.
+            unknown: void,
+            /// The peer is a client with the given id.
+            client: u128,
+            /// The peer is a replica with the given id.
+            replica: u16,
+        } = .none,
+
+        /// This completion is used for all recv operations.
+        recv_completion: IO.Completion = undefined,
+        /// Number of bytes of the current header/message that have already been recieved.
+        bytes_received: usize = 0,
+        incoming_header: Header = undefined,
+        incoming_message: *Message = undefined,
+
+        /// This completion is used for all send operations.
+        send_completion: IO.Completion = undefined,
+        /// The queue of messages to send to the client.
+        /// Empty unless peer == .client
+        send_queue: FIFO(Envelope) = .{},
+        bytes_sent: usize = 0,
+
+        /// Given a newly accepted fd, start receiving messages on it.
+        /// Callbacks will be continously re-registered until close() is
+        /// called and the connection is terminated.
+        fn receive_messages(self: *Connection, fd: os.socket_t) void {
+            assert(self.fd == -1);
+            assert(self.peer == .none);
+            self.fd = fd;
+            self.peer = unknown;
+            self.recv_header();
+        }
+
+        fn recv_header(self: *Connection) void {
+            // The fd may have been closed due to an error while sending.
+            if (self.fd == -1) return;
+            self.message_bus.io.recv(
+                *ReplicaConnection,
+                self,
+                complete_recv_header,
+                &self.completion,
+                mem.asBytes(&self.incoming_header)[self.bytes_received..],
+                os.MSG_NOSIGNAL,
+            );
+        }
+
+        fn complete_recv_header(self: *Connection, completion: *IO.Completion, result: RecvError!usize) void {
+            self.bytes_received += result catch |err| {
+                // TODO: maybe don't need to close on *every* error
+                log.err("error receiving data from {}: {}", .{ self.peer, err });
+                self.close();
+                return;
+            };
+
+            if (self.bytes_received < @sizeOf(Header)) {
+                // The header has not yet been fully received.
+                self.recv_header();
+                return;
+            }
+            assert(self.bytes_received == @sizeOf(Header));
+
+            if (!self.incoming_header.valid_checksum()) {
```

Comment:
We should also pass the `cluster_id` to MessageBus at `init()` so that we can close connections early at the first sign of a header addressed to the wrong cluster.

Reply:
Alternatively, we could also read it off `self.server.cluster` (and do the same with `self.server.replica`)?

Reply:
Reading off of `self.server` for both of these makes sense to me.

Reply:
Great, I saw the latest revision accepts `server_index` in `MessageBus.init()`. Was your reasoning behind this because the MessageBus and Replica are cyclically referenced and one may init before the other (or in any order)? i.e. We need to pass `server_index` explicitly because `Replica.init()` may not yet have completed and `self.server.replica` may not yet be initialized?

Reply:
Yep, that was the thought although it doesn't seem like `Replica.init()` depends on the passed `MessageBus` being initialized currently so we could potentially avoid the explicitness. This stuff isn't hard to tweak down the line as we actually start fitting the pieces together though.